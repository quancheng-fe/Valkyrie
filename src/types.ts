import Koa from 'koa'
export type loadFunction<Config, ReturnType> = (
  path: string,
  app: Koa,
  config?: Config
) => Promise<ReturnType> | ReturnType
