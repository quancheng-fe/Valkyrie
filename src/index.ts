export * from 'typeorm-typedi-extensions'
export * from 'type-graphql'
export {
  ContainerInstance,
  Container,
  Inject,
  InjectMany,
  Service,
  Token
} from 'typedi'
export { Repository } from 'typeorm'
export { Context, Middleware } from 'koa'
export { createServer } from './server'
export { InjectGrpcService } from './lib/loadGrpc'
export * from './types'
