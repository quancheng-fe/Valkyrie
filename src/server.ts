import http from 'http'
import Koa from 'koa'
import pino, { Logger } from 'pino'
import { config as envConfig } from 'dotenv'
import { Config } from 'apollo-server-koa'
import Redis from 'ioredis'

import { loadGraphql } from './lib/loadGraphql'
import { loadMiddlewares } from './lib/loadMiddleware'
import { loadPlugin } from './lib/loadPlugin'
import { loadConfigFromACM, NodeBaseConfig } from './lib/loadConfig'
import { loadORM } from './lib/loadORM'
import { loadNext } from './lib/loadNext'
import { loadSentry } from './lib/loadSentry'
import { setupEureka } from './lib/loadGrpc'
import { gracefulShutDown } from './lib/gracefulShutDown'
import { errorHandler } from './middleware/errhandler'
import { zipkinTracer } from './middleware/tracer'

export interface ServerOptions extends Partial<NodeBaseConfig> {
  name: string
  root: string
  graphqlServer: Config
  onShuttingDown?: Function
}

export const createServer = async (
  conf: ServerOptions,
  app?: Koa & Record<string, any> & { logger?: Logger }
): Promise<http.Server> => {
  envConfig()

  app = app || new Koa()

  const {
    SENTRY_DSN,
    ACM_KEY,
    ACM_ENDPOINT,
    ACM_NAMESPACE,
    ACM_SECRET,
    LOG_LEVEL,
    NODE_ENV
  } = process.env

  // init Logger
  const logger = pino({
    level: LOG_LEVEL || 'info',
    prettyPrint: NODE_ENV !== 'production'
  })

  logger.debug('env: ', process.env)

  app.logger = logger

  // init loadPlugin factory function
  const load = loadPlugin(app)
  load('logger', logger)
  load('sentry', loadSentry(SENTRY_DSN))

  app.use(errorHandler)

  // load config from aliyun acm
  const config = await loadConfigFromACM({
    accessKey: ACM_KEY,
    secretKey: ACM_SECRET,
    endpoint: ACM_ENDPOINT,
    namespace: ACM_NAMESPACE
  })

  // merge local and remote config
  const appConfig = { ...config, ...conf }
  logger.info('config: ', appConfig)

  const { root, redis, zipkin } = appConfig

  app.use(zipkinTracer(zipkin))

  // load internal plugin
  load('redis', new Redis(redis))

  // load orm entities from files named '**.entity.**'
  load('connection', await loadORM(root, app, appConfig))

  // load middlewares from files named '**.middleware.**'
  await loadMiddlewares(root, app)

  // load middlewares from files named '**.resolver.**'
  await loadGraphql(root, app, appConfig)

  const eurekaClient = await setupEureka(appConfig, logger)

  await loadNext(root, app)

  // print error message
  app.on('error', err => {
    logger.error('server error', err)
  })

  const server = http.createServer(app.callback())

  // graceful shutdown
  process.on(
    'SIGTERM',
    gracefulShutDown(server, logger, {
      onShutDown() {
        eurekaClient.stop()
      }
    })
  )

  return server
}
