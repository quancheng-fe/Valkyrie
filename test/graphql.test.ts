import 'reflect-metadata'
import { resolve } from 'path'
import request, { SuperTest, Test } from 'supertest'
import { createServer } from '../src'

let agent: SuperTest<Test>

beforeAll(async () => {
  const server = await createServer({
    name: 'test-app',
    root: resolve(__dirname, './fixture/app'),
    graphqlServer: {
      introspection: true,
      playground: true
    }
  })
  agent = request.agent(server)
})

const TestQueryDocument = `
  query {
    recipe {
      id
      title
      description
    }
  }
`

it('query will work', async () => {
  const response = await agent.post('/graphql').send({
    operationName: null,
    query: TestQueryDocument
  })

  expect(response.status).toBe(200)
  expect(response.body).toMatchSnapshot('graphql-test-query-recipe')
})
