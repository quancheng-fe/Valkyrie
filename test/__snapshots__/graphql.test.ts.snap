// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query will work: graphql-test-query-recipe 1`] = `
Object {
  "data": Object {
    "recipe": Object {
      "description": "test",
      "id": "1",
      "title": "test",
    },
  },
}
`;
